syntax = "proto3";

import "tensorflow/core/framework/types.proto";

package ml;

enum CombinerType {
  COMBINER_TYPE_UNKNOWN = 0;
  COMBINER_TYPE_MEAN = 1;
  COMBINER_TYPE_SQRT = 2;
  COMBINER_TYPE_SUM = 3;
}

message Bucket {
  float upper_boundary = 1;
  int64 sample_count = 2;
}

message Histogram { repeated Bucket buckets = 1; }

message NumericStats {
  float min = 1;
  float max = 2;
  float mean = 3;
  float std = 4;
  Histogram histogram = 5;
}

message RealValuedColumn {
  int32 dimension = 1;
  tensorflow.DataType dtype = 2;
}

message IntegerizedSparseColumn {
  int64 bucket_size = 1;
  CombinerType combiner = 2;
  tensorflow.DataType dtype = 3;
}

message HashBucketSparseColumn {
  int64 hash_bucket_size = 1;
  CombinerType combiner = 2;
  tensorflow.DataType dtype = 3;
}

// Converts real valued features to integer features.
// The bucketizing boundaries are found in the stats
// of the source column.
message BucketizedColumn {}

message OneHotColumn {}

message EmbeddingColumn {
  int32 dimension = 1;
  CombinerType combiner = 2;
}

message SharedEmbeddingColumn {
  int32 dimension = 1;
  CombinerType combiner = 2;
}

message FeatureColumnConfig {
  string name = 1;
  oneof specific_column {
    BucketizedColumn bucketized = 2;
    EmbeddingColumn embedding = 3;
    HashBucketSparseColumn hash_bucket = 4;
    IntegerizedSparseColumn integerized = 5;
    OneHotColumn one_hot = 6;
    RealValuedColumn real_valued = 7;
    SharedEmbeddingColumn shared_embedding = 8;
  }
  // Names of the columns this feature column depends on.
  repeated string dependency = 9;
  bool include = 10;
}

message FeatureColumnConfigSchema {
  string description = 1;
  repeated FeatureColumnConfig column_config = 2;
}
